dwave/optimization/include/dwave-optimization/array.hpp

```diff
@@ -413,17 +435,40 @@ class ArrayIterator {
             return offset;
         }
 
-        std::ptrdiff_t advance(std::ptrdiff_t n) {
-            if (n < 0) {
-                assert(false && "not implemented yet");
-                unreachable();
+        std::ptrdiff_t unadvance() {
+            std::ptrdiff_t offset = 0;
+
+            ssize_t dim = ndim - 1;
+            for (; dim >= 1; --dim) {
+                offset -= strides[dim];
+
+                if (--loc[dim - 1] >= 0) break;
+
+                assert(loc[dim - 1] == -1);
+
+                offset += strides[dim] * shape[dim];
+                loc[dim - 1] = shape[dim] - 1;
+            }
+            if (dim == 0) {
+                offset -= strides[dim];
             }
 
+            return offset;
+        }
+
+        std::ptrdiff_t advance(std::ptrdiff_t n) {
+            std::ptrdiff_t offset = 0;
+
             // do the dumb thing for now - we can improve the performance of this
             // later
-            std::ptrdiff_t offset = 0;
-            for (std::ptrdiff_t i = 0; i < n; ++i) {
-                offset += advance();
+            if (n < 0) {
+                for (std::ptrdiff_t i = 0; i > n; --i) {
+                    offset += unadvance();
+                }
+            } else {
+                for (std::ptrdiff_t i = 0; i < n; ++i) {
+                    offset += advance();
+                }
             }
             return offset;
         }
```

Comment:
I think we could also just do
```c++
        std::ptrdiff_t advance(std::ptrdiff_t n) {
            std::ptrdiff_t offset = 0;

            for (; n > 0; --n) offset += unadvance();
            for (; n < 0; ++n) offset += advance();

            return offset;
        }
```
likely after the compiler has at it, it won't make much difference. But I guess this is a few fewer instructions/memory locations if it's taken literally.

Reply:
Done, though I had to switch them :stuck_out_tongue: